scripts/bluetooth/mesh/mesh_dfu_metadata.py

```diff
@@ -0,0 +1,413 @@
+                    if not vnd:
+                        elem_item.sig_model_add(id1)
+                    else:
+                        elem_item.vnd_model_add(id1, id2)
+
+            if sig_count > 0:
+                models_unpack(sig_ptr, elem_item, False)
+            if vnd_count > 0:
+                models_unpack(vnd_ptr, elem_item, True)
+        return comp
+
+def parse_comp_data(elf_path, kconfigs):
+    """
+    Extract composition data from .elf and .config file.
+
+    Parameters:
+        kconfigs (KConfig): A KCoonfig object representing Kconfig options used for the firmware to compile with
+    Returns:
+        Parsed Composition data
+    """
+    try:
+        addr = find_comp_data_from_dwarf(elf_path)
+        return read_comp_data(elf_path, addr, kconfigs)
+    except Exception:
+        traceback.print_exc()
+        exit_with_error_msg(f"Failed to extract composition data from .elf file")
+
+def encoded_metadata_get(version, comp):
+    core_type = 0 # FIX ME: For now, hardcoded to application core
+    binary_size = os.path.getsize(os.path.join(bin_path, 'app_update.bin'))
+    elem_cnt = len(comp.elems)
+
+    bytestring = bytearray()
+    bytestring.append(version["major"])
+    bytestring.append(version["minor"])
+    bytestring.extend(version["revision"].to_bytes(2, 'little'))
+    bytestring.extend(version["build_number"].to_bytes(4, 'little'))
+    bytestring.extend(binary_size.to_bytes(3, 'little'))
+    bytestring.append(core_type)
+    bytestring.extend(comp.hash_generate().to_bytes(4, 'little'))
+    bytestring.extend(elem_cnt.to_bytes(2, 'little'))
+    # FIX ME: Support user data
+    return bytestring
+
+if __name__ == "__main__":
+    if len(sys.argv) != 2:
+        exit_with_error_msg("Unexpected number of script input arguments")
+
+    bin_path = sys.argv[1]
+    zip_path = os.path.abspath(os.path.join(bin_path, 'dfu_application.zip'))
+    metadata_path = os.path.abspath(os.path.join(bin_path, FILE_NAME))
+    elf_path = os.path.abspath(os.path.join(bin_path, 'zephyr.elf'))
+    config_path = os.path.abspath(os.path.join(bin_path, '.config'))
+
+    zip = ZipFile(zip_path, "a")
+    if FILE_NAME_IN_ZIP in zip.namelist():
+        # Mesh metadata already present in zip file
+        sys.exit(0)
+
+    kconfigs = KConfig.from_file(config_path)
+    comp = parse_comp_data(elf_path, kconfigs)
+    version = kconfigs.version_parse()
+    encoded_metadata = encoded_metadata_get(version, comp)
+
+    j_dict = {
+        "sign_version": version,
+        "composition_data": comp.dict_generate(),
+        "composition_hash": str(hex(comp.hash_generate())),
+        "encoded_metadata": str(encoded_metadata.hex()),
+    }
+
+    with open(metadata_path, "w") as outfile:
+        outfile.write(json.dumps(j_dict, indent=4))
+    zip.write(metadata_path, FILE_NAME_IN_ZIP)
+    zip.close()
+
+    print("Bluetooth Mesh Composition metadata generated:")
+    print(f"\tMetadata hash: {j_dict['composition_hash']}")
```

Comment:
Sorry for nitpicking, but this hash has no sense on its own without other parameters. The encoded metadata is what we really need to be printed.